### PR TITLE
Tyler Starr Project 05 - Vertex Shader

### PR DESCRIPTION
**Pull request by sta14031**
screenshot

This project took considerably more effort than the previous ones and we had to collaborate to solve some of the problems that we encountered. Thanks to everybody who showed up to our meeting on Wednesday night.

### CODE REVIEW

**Comment by sta14031** [comment]
I also copied Brother Christensen's refactored Attributes class in this commit. I see the utility of structuring the class in the way that he did, the idea just didn't come to me at the time. I tried to emulate that kind of modular design with the Matrix class and the transformation helper functions.

File: definitions.h [hunk 1]
```diff
@@ -214,37 +221,234 @@ class BufferImage : public Buffer2D<PIXEL>
...
+    attr.matrix[0]  = v.x;
+    attr.matrix[5]  = v.y;
+    attr.matrix[10] = v.z;
+}
+
+// This function takes an angle in degrees and converts it to radians
+void rotateZMatrix(Attributes& attr, double angle) {
+    angle = angle * M_PI / 180.0;
+    // compute trig functions here so we dont have to do them twice
+    double sinangle = sin(angle);
```
Comment (G-Slim) [inline definitions.h]: Great job pre-computing values. Your code is looking very efficient!

File: definitions.h [hunk 2]
```diff
@@ -214,37 +221,234 @@ class BufferImage : public Buffer2D<PIXEL>
...
+    double cosangle = cos(angle);
+
+    attr.matrix[0]  = cosangle;
+    attr.matrix[2]  = sinangle;
+    attr.matrix[8]  = -sinangle;
+    attr.matrix[10] = cosangle;
+}
+
+// Does all three previous operations at once in right to left order
+void STRMatrix(Attributes& attr, Vertex s, Vertex t, double r) {
```
Comment (G-Slim) [inline definitions.h]: 4x4 matrix multiplication is looking good. This looks like a great way to perform all of the transformations without making things too complicated. Good job.

**Comment by G-Dubs** [comment]
This looks pretty good.  All the triangles are in the right spot & have the right rotation.  You must have remembered all your return statements.

**Pull request merged**